### ISSUE TITLE
Add incremental strain to small strain models

### ISSUE THREAD

**Issue by dschwen** [idaholab/moose#6210]
This would allow users to use small strain formulations with stress classes that require incremental strain formulations.

We could either add a parameter `_incremental_form(getParam<bool>("incremental_form"))` to `ComputeSmallStrain` like in `ComputeStressFreeStrainBase`, or we could derive from `ComputeSmallStrain`.

Ping @tonkmr @bwspenc

**Comment by bwspenc** [comment]
The motivation behind this is to make the models more "plug and play".  Currently, there are separate versions of some of the stress calculator models to work with small strain and with finite strain. A big motivation behind the current system is to avoid having a separate version of each model for each combination of elasticity tensor and strain calculation. The fact that we have large and small strain variants of models shows that we haven't achieved that goal.

**Comment by tonkmr** [comment]
@cpritam and I talked about this a lot before I left and we almost switched everything to an incremental form. What kept us from doing it is the fact that you have to have a stateful material property tensor that is the old strain. Especially at the microscale we do a lot of simulations that are pure linear elastic, and it seemed like a high cost to have the extra storage be required for all models just for convenience, even if they don't need it. Our simulations are already very memory intensive in 3D and I would hate to make it worse when it isn't necessary.

So, my thought was that the ONLY small strain, non-incremental form we would have would be linear elasticity, and everything else would be incremental. That way, we could minimize the number of models that aren't plug and play.

**Comment by tonkmr** [comment]
Let me clarify that if you can figure out a way to make ComputeSmallStrain work for both total strain and incremental strain, I would be very supportive. We just need to make sure we have a good way to only require the stateful old strain with the incremental form.

**Comment by bwspenc** [comment]
@tonkmr I agree that we want to have the ability to do linear elastic calculations with no state. We'll think about ideas to accomplish all of the goals here.
